alibi_detect/utils/pytorch/kernels.py

```diff
@@ -14,22 +18,76 @@ def __init__(self, sigma: Optional[torch.Tensor] = None) -> None:
         Parameters
         ----------
         sigma
-            Optional sigma used for the kernel.
+            Bandwidth used for the kernel. Needn't be specified if being inferred or trained.
+            Can pass multiple values to eval kernel with and then average.
+        trainable
+            Whether or not to track gradients w.r.t. sigma to allow it to be trained.
         """
         super().__init__()
-        self.sigma = sigma
+        if sigma is None:
+            self.log_sigma = nn.Parameter(torch.empty(1), requires_grad=trainable)
+            self.init_required = True  # TODO: Is this right now?
+        else:
+            sigma = sigma.reshape(-1)  # [Ns,]
+            self.log_sigma = nn.Parameter(sigma.log(), requires_grad=trainable)
+            self.init_required = False
+        self.trainable = trainable
 
-    def __call__(self, x: torch.Tensor, y: torch.Tensor, infer_sigma: bool = False) -> torch.Tensor:
+    @property
+    def sigma(self) -> torch.Tensor:
+        return self.log_sigma.exp()
 
-        dist = distance.squared_pairwise_distance(x, y)  # [Nx, Ny]
+    def forward(self, x: torch.Tensor, y: torch.Tensor, infer_sigma: bool = False) -> torch.Tensor:
 
-        if infer_sigma:
+        dist = distance.squared_pairwise_distance(x.flatten(1), y.flatten(1))  # [Nx, Ny]
+
+        if infer_sigma or self.init_required:
+            if self.trainable and infer_sigma:
+                raise ValueError("Gradients cannot be computed w.r.t. an inferred sigma value")
             n = min(x.shape[0], y.shape[0])
             n = n if (x[:n] == y[:n]).all() and x.shape == y.shape else 0
             n_median = n + (np.prod(dist.shape) - n) // 2 - 1
-            self.sigma = (.5 * dist.flatten().sort().values[n_median].unsqueeze(dim=-1)) ** .5
+            sigma = (.5 * dist.flatten().sort().values[n_median].unsqueeze(dim=-1)) ** .5
+            with torch.no_grad():
+                self.log_sigma.copy_(sigma.log().clone())
+            self.init_required = False
 
         gamma = 1. / (2. * self.sigma ** 2)   # [Ns,]
         # TODO: do matrix multiplication after all?
         kernel_mat = torch.exp(- torch.cat([(g * dist)[None, :, :] for g in gamma], dim=0))  # [Ns, Nx, Ny]
         return kernel_mat.mean(dim=0)  # [Nx, Ny]
+
+
+class DeepKernel(nn.Module):
+    """"
+    Computes simmilarities as k(x,y) = (1-eps)*k_a(proj(x), proj(y)) + eps*k_b(x,y)
+    """
+    def __init__(
+        self,
+        proj: nn.Module,
+        kernel_a: nn.Module = GaussianRBF(trainable=True),
+        kernel_b: nn.Module = GaussianRBF(trainable=True),
+        eps: Union[float, str] = 'trainable'
```

Comment:
Is there a reason why we don't choose for 2 kwargs: `eps` and `eps_trainable` or something similar and instead set `eps` in the trainable case by default to .5?

Reply:
I guess we could. I thought this was less ambiguous as it is either trainable or the value provided, it can't be both.

And `logit_eps=0` is same as `eps=0.5`.

Reply:
Yes, my point was more that now if you set it to 'trainable' `eps` is by default 0.5 but I am not sure if that's always a good setting in the trade off between the `kernel_a` and `kernel_b`.

Reply:
Ah, I don't think it's a problem given that it's trainable. Here the init won't cause a 0 gradient problem like it does for `sigma` in `GaussianRBF`.